### PR TITLE
Unify TLS destructor list implementations

### PR DESCRIPTION
**Pull request by joboet**
Part of #110897.

Currently, nearly every platform implements its own TLS destructor list. This adds unnecessary code and means that issues like #116390 are harder to find and fix. Therefore, this PR unifies all of these implementations into a shared one inside `sys::common::fast_local`, leaving behind only the platform calls needed to ensure the list is emptied.

This changes behaviour on some Linux-like platforms, because we now keep our own list. This should however not impact performance too much, as the platform calls would also have needed to allocate.

Because it is only used on UNIX and is now much easier, I've removed the fallback implementation in `sys_common`. Because `StaticKey` may now be unused, this resulted in errors for Windows, which is why I have refactored that code to move the at-exit hack to the new `thread_local_guard` module. If it causes too much review pain, I can split this part out.

Tested on aarch64 macOS, checked on all other platforms. Best reviewed per-commit.

@rustbot label +T-libs +A-thread-locals

### CODE REVIEW

File: library/std/src/sys/hermit/thread_local_dtor.rs
Comment (RalfJung) [inline library/std/src/sys/hermit/thread_local_dtor.rs:1]: Why is this file called "dtor" but the other similar file have been renamed to "guard"?

Comment (joboet) [reply library/std/src/sys/hermit/thread_local_dtor.rs:1]: That was totally intentional, I wanted to add a bit more personality to the module (/s) 😁

**Comment by joboet** [comment]
This makes the `dylib` unloading problem described [here](https://github.com/rust-lang/rust/issues/110897#issuecomment-1577134152) even worse. I'll get some other things done before coming back to this.

@rustbot author

**Comment by joboet** [comment]
Closing in favour of #126523.

**Pull request closed without merging**